File: caid/impl/js/run-vectors.mjs
// run-vectors.mjs - runs every vector in conformance/vectors.json against
// caid.mjs. Prints pass/fail per vector and exits nonzero on any failure.
//
// Usage: node run-vectors.mjs
//
// Vector kinds:
//   compute - computeCaid(input.object, {suite: input.suite, definitions})
//   verify  - verifyCaid(input.object, input.caid, {definitions})
//   parse   - parseCaid(input.caid)
//
// Optional per-vector "relation" cross-checks:
//   same_caid_as       - actual computed caid must equal that vector's
//   different_caid_from - actual computed caid must differ from that vector's

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { computeCaid, verifyCaid, parseCaid } from "./caid.mjs";

const here = dirname(fileURLToPath(import.meta.url));
const vectorsPath = join(here, "..", "..", "conformance", "vectors.json");
const suite = JSON.parse(readFileSync(vectorsPath, "utf8"));

function deepEqual(a, b) {
  return JSON.stringify(a) === JSON.stringify(b);
}

let pass = 0;
let fail = 0;
const actualCaids = new Map(); // id -> caid string (successful computes only)

function report(id, ok, detail) {
  if (ok) {
    pass++;
    console.log("PASS " + id);
  } else {
    fail++;
    console.log("FAIL " + id);
    if (detail) console.log("     " + detail);
  }
}

for (const v of suite.vectors) {
  let actual;
  if (v.kind === "compute") {
    actual = computeCaid(v.input.object, {
      suite: v.input.suite,
      definitions: v.definitions,
    });
    if (actual && typeof actual.caid === "string") {
      actualCaids.set(v.id, actual.caid);
    }
  } else if (v.kind === "verify") {
    actual = verifyCaid(v.input.object, v.input.caid, {
      definitions: v.definitions,
    });
  } else if (v.kind === "parse") {
    actual = parseCaid(v.input.caid);
  } else {
    report(v.id, false, "unknown vector kind: " + v.kind);
    continue;
  }
  const ok = deepEqual(actual, v.expect);
  report(
    v.id,
    ok,
    ok
      ? null
      : "expected " + JSON.stringify(v.expect) + " got " + JSON.stringify(actual)
  );
}

// Relation cross-checks over ACTUAL computed values, not just the
// expectations written in the file.
for (const v of suite.vectors) {
  if (!v.relation) continue;
  const targetId = v.relation.same_caid_as || v.relation.different_caid_from;
  const mine = actualCaids.get(v.id);
  const theirs = actualCaids.get(targetId);
  if (mine === undefined || theirs === undefined) {
    report(v.id + " (relation)", false, "missing computed caid for relation");
    continue;
  }
  if (v.relation.same_caid_as) {
    report(
      v.id + " same_caid_as " + targetId,
      mine === theirs,
      mine === theirs ? null : mine + " != " + theirs
    );
  } else {
    report(
      v.id + " different_caid_from " + targetId,
      mine !== theirs,
      mine !== theirs ? null : "caids unexpectedly equal: " + mine
    );
  }
}

console.log("");
console.log(pass + " passed, " + fail + " failed, " + suite.vectors.length + " vectors");
process.exit(fail > 0 ? 1 : 0);

File: caid/impl/python/run_vectors.py
# run_vectors.py - runs every vector in conformance/vectors.json against
# caid.py. Prints pass/fail per vector and exits nonzero on any failure.
#
# Usage: python3 run_vectors.py
#
# Vector kinds:
#   compute - compute_caid(input.object, {suite: input.suite, definitions})
#   verify  - verify_caid(input.object, input.caid, {definitions})
#   parse   - parse_caid(input.caid)
#
# Optional per-vector "relation" cross-checks:
#   same_caid_as        - actual computed caid must equal that vector's
#   different_caid_from - actual computed caid must differ from that vector's

import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from caid import compute_caid, verify_caid, parse_caid  # noqa: E402

HERE = os.path.dirname(os.path.abspath(__file__))
VECTORS_PATH = os.path.join(HERE, "..", "..", "conformance", "vectors.json")

with open(VECTORS_PATH, "r", encoding="utf-8") as fh:
    suite = json.load(fh)


def deep_equal(a, b):
    return json.dumps(a, sort_keys=True) == json.dumps(b, sort_keys=True)


counts = {"pass": 0, "fail": 0}
actual_caids = {}  # id -> caid string (successful computes only)


def report(vector_id, ok, detail=None):
    if ok:
        counts["pass"] += 1
        print("PASS " + vector_id)
    else:
        counts["fail"] += 1
        print("FAIL " + vector_id)
        if detail:
            print("     " + detail)


for v in suite["vectors"]:
    kind = v.get("kind")
    if kind == "compute":
        actual = compute_caid(
            v["input"]["object"],
            {"suite": v["input"].get("suite"), "definitions": v["definitions"]},
        )
        if isinstance(actual, dict) and isinstance(actual.get("caid"), str):
            actual_caids[v["id"]] = actual["caid"]
    elif kind == "verify":
        actual = verify_caid(
            v["input"]["object"],
            v["input"]["caid"],
            {"definitions": v["definitions"]},
        )
    elif kind == "parse":
        actual = parse_caid(v["input"]["caid"])
    else:
        report(v["id"], False, "unknown vector kind: " + str(kind))
        continue
    ok = deep_equal(actual, v["expect"])
    report(
        v["id"],
        ok,
        None
        if ok
        else "expected "
        + json.dumps(v["expect"])
        + " got "
        + json.dumps(actual),
    )

# Relation cross-checks over ACTUAL computed values, not just the
# expectations written in the file.
for v in suite["vectors"]:
    relation = v.get("relation")
    if not relation:
        continue
    target_id = relation.get("same_caid_as") or relation.get("different_caid_from")
    mine = actual_caids.get(v["id"])
    theirs = actual_caids.get(target_id)
    if mine is None or theirs is None:
        report(v["id"] + " (relation)", False, "missing computed caid for relation")
        continue
    if relation.get("same_caid_as"):
        report(
            v["id"] + " same_caid_as " + target_id,
            mine == theirs,
            None if mine == theirs else mine + " != " + theirs,
        )
    else:
        report(
            v["id"] + " different_caid_from " + target_id,
            mine != theirs,
            None if mine != theirs else "caids unexpectedly equal: " + mine,
        )

print("")
print(
    "%d passed, %d failed, %d vectors"
    % (counts["pass"], counts["fail"], len(suite["vectors"]))
)
sys.exit(1 if counts["fail"] > 0 else 0)
